### PR TITLE
Inserts latency profiling timers at various interesting points.

### PR DESCRIPTION
**Pull request by vjpai**
This is just for inspection and discussion. It needs a lot of effort to improve, though it is already somewhat usable at the small scale. For example, we're not properly separating out per-call state through the stack yet, which could either be done through a thread-local or actually passing a value up and down the stack. For now, then, it's not bad for looking at proto and system call overhead in a contentionless RPC case.

### CODE REVIEW

File: src/core/iomgr/tcp_posix.c [hunk 1]
```diff
@@ -402,6 +406,7 @@ static void grpc_tcp_handle_read(void *arg /* grpc_tcp */, int success) {
       ++iov_size;
     }
   }
+  GRPC_TIMER_MARK(HANDLE_READ_BEGIN, 0);
```
Comment (ctiller) [inline src/core/iomgr/tcp_posix.c]: aliasing this name

File: src/cpp/server/async_server_context.cc [hunk 2]
```diff
@@ -87,7 +89,9 @@ bool AsyncServerContext::StartWriteStatus(const Status& status) {
 
 bool AsyncServerContext::ParseRead(grpc_byte_buffer* read_buffer) {
   GPR_ASSERT(request_);
+  GRPC_TIMER_MARK(DESER_PROTO_BEGIN, call_->call());
   bool success = DeserializeProto(read_buffer, request_);
+  GRPC_TIMER_MARK(DESER_PROTO_BEGIN, call_->call());
```
Comment (ctiller) [inline src/cpp/server/async_server_context.cc]: _END

**Comment by ctiller** [comment]
This is already wildly useful!

Thanks Vijay

**Comment by ctiller** [comment]
As an aside... we might want to tweak the api soon, exposing:

GRPC_TIMER_BEGIN(name, object)
GRPC_TIMER_END(name)

and start capturing the stack with a thread local.

**Comment by ctiller** [comment]
If we can fix those two names, I'd be happy to merge this.

**Comment by vjpai** [comment]
Thanks for the feedback. I'll be at my desk shortly and will get those changes made.

File: src/core/iomgr/tcp_posix.c (hunk 1 above)
Comment (vjpai) [inline src/core/iomgr/tcp_posix.c]: done.

File: src/cpp/server/async_server_context.cc (hunk 2 above)
Comment (vjpai) [inline src/cpp/server/async_server_context.cc]: done

**Comment by vjpai** [comment]
That escalated quickly. I've stripped the DO NOT MERGE branch and Travis is green; feel free to merge. Thanks!

**Pull request merged**